### PR TITLE
Jwt

### PR DESCRIPTION
**Pull request by AgroFix**

### CODE REVIEW

File: src/main/java/mate/academy/bookstore/security/CustomUserDetailsService.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+import mate.academy.bookstore.repository.UserRepository;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.security.core.userdetails.UsernameNotFoundException;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class CustomUserDetailsService implements UserDetailsService {
+    private static final String USER_NOT_FOUND_BY_EMAIL_MSG = "Can't find user by email: ";
```
Comment (andrii-hoienko) [inline src/main/java/mate/academy/bookstore/security/CustomUserDetailsService.java]:
```suggestion
    private static final String USER_NOT_FOUND_BY_EMAIL_MSG = "Can't find user by email: %s";
```

File: src/main/java/mate/academy/bookstore/security/CustomUserDetailsService.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    private final UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String email)
+            throws UsernameNotFoundException {
+        return userRepository.findByEmail(email).orElseThrow(
+                        () -> new UsernameNotFoundException(USER_NOT_FOUND_BY_EMAIL_MSG + email));
```
Comment (andrii-hoienko) [inline src/main/java/mate/academy/bookstore/security/CustomUserDetailsService.java]: Better to use String.format in such cases

File: src/main/java/mate/academy/bookstore/security/JwtAuthenticationFilter.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+            SecurityContextHolder.getContext().setAuthentication(authentication);
+        }
+
+        filterChain.doFilter(request, response);
+    }
+
+    private Optional<String> getToken(HttpServletRequest request) {
+        String bearerToken = request.getHeader("Authorization");
+        if (StringUtils.hasText(bearerToken) && bearerToken.startsWith("Bearer ")) {
+            return Optional.of(bearerToken.substring(7));
```
Comment (andrii-hoienko) [inline src/main/java/mate/academy/bookstore/security/JwtAuthenticationFilter.java]: Magic number

File: src/main/java/mate/academy/bookstore/security/JwtAuthenticationFilter.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+            );
+            SecurityContextHolder.getContext().setAuthentication(authentication);
+        }
+
+        filterChain.doFilter(request, response);
+    }
+
+    private Optional<String> getToken(HttpServletRequest request) {
+        String bearerToken = request.getHeader("Authorization");
+        if (StringUtils.hasText(bearerToken) && bearerToken.startsWith("Bearer ")) {
```
Comment (andrii-hoienko) [inline src/main/java/mate/academy/bookstore/security/JwtAuthenticationFilter.java]: Please make "Bearer " constant

File: src/main/java/mate/academy/bookstore/security/JwtUtil.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    public boolean isValidToken(String token) {
+        try {
+            Jws<Claims> claimsJws = Jwts.parser()
+                    .verifyWith(secret)
+                    .build()
+                    .parseSignedClaims(token);
+
+            return !claimsJws.getPayload().getExpiration().before(new Date());
+        } catch (JwtException | IllegalArgumentException e) {
+            throw new JwtException("Expired or invalid JWT token");
```
Comment (andrii-hoienko) [inline src/main/java/mate/academy/bookstore/security/JwtUtil.java]: Please be consistent, in some cases you extract error messages in constants in some cases not, please decide on behavior with error messages and apply it to all project

File: src/main/java/mate/academy/bookstore/model/User.java [hunk 6]
```diff
@@ -24,4 +36,49 @@ public class User {
     @Column(nullable = false)
     private String lastName;
     private String shippingAddress;
+    @ManyToMany(fetch = FetchType.EAGER)
+    @JoinTable(
+            name = "users_roles",
+            joinColumns = {@JoinColumn(name = "user_id")},
+            inverseJoinColumns = {@JoinColumn(name = "role_id")}
+    )
+    private Set<Role> roles;
```
Comment (andrii-hoienko) [inline src/main/java/mate/academy/bookstore/model/User.java]: Please exclude relations from equals/hashCode and toString

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]

File: src/main/resources/db/changelog/changes/08-insert-users-roles.yaml
Comment (d1sam) [inline src/main/resources/db/changelog/changes/08-insert-users-roles.yaml:1]: it's better to insert user by valueComputed and preConditions, example:
 ```
     preConditions:
        - sqlCheck:
            expectedResult: 1
            sql: SELECT COUNT(*) FROM users WHERE email='admin@example.com'
        - sqlCheck:
            expectedResult: 1
            sql: SELECT COUNT(*) FROM roles WHERE name='ROLE_ADMIN'
      changes:
        - insert:
            tableName: user_role
            columns:
              - column:
                  name: user_id
                  valueComputed: (SELECT id FROM users WHERE email='admin@example.com')
              - column:
                  name: role_id
                  valueComputed: (SELECT id FROM roles WHERE name='ROLE_ADMIN')
```

File: src/main/resources/db/changelog/changes/06-create-users-roles-table.yaml
Comment (d1sam) [inline src/main/resources/db/changelog/changes/06-create-users-roles-table.yaml:1]: users_roles should have composite key.

File: src/main/java/mate/academy/bookstore/model/User.java [hunk 7]
```diff
@@ -24,4 +40,49 @@ public class User {
...
+    }
+
+    @Override
+    public String getPassword() {
+        return password;
+    }
+
+    @Override
+    public boolean isAccountNonExpired() {
+        return true;
```
Comment (d1sam) [inline src/main/java/mate/academy/bookstore/model/User.java]:
```suggestion
        return !isDeleted;
```

File: src/main/java/mate/academy/bookstore/model/User.java [hunk 8]
```diff
@@ -24,4 +40,49 @@ public class User {
...
+    }
+
+    @Override
+    public boolean isAccountNonLocked() {
+        return true;
```
Comment (d1sam) [inline src/main/java/mate/academy/bookstore/model/User.java]:
```suggestion
        return !isDeleted;
```

File: src/main/java/mate/academy/bookstore/model/User.java [hunk 9]
```diff
@@ -24,4 +40,49 @@ public class User {
...
+    }
+
+    @Override
+    public boolean isCredentialsNonExpired() {
+        return true;
```
Comment (d1sam) [inline src/main/java/mate/academy/bookstore/model/User.java]:
```suggestion
        return !isDeleted;
```

**Review by d1sam — CHANGES_REQUESTED** [review]
Good job, but some changes should be done and why I see changes from previous prs in this pr?

File: src/main/java/mate/academy/bookstore/dto/UserLoginRequestDto.java [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+import jakarta.validation.constraints.Email;
+import jakarta.validation.constraints.NotBlank;
+import jakarta.validation.constraints.Size;
+import lombok.Data;
+
+@Data
+public class UserLoginRequestDto {
+    @NotBlank
+    @Email
+    private String email;
```
Comment (vovasalyha) [inline src/main/java/mate/academy/bookstore/dto/UserLoginRequestDto.java:12]: Validate also maximum size for email, to not exceed DB schema constraints.

**Review by vovasalyha — APPROVED** [review]

**Pull request merged**